File: 16.最接近的三数之和.cs
/*
 * @lc app=leetcode.cn id=16 lang=csharp
 *
 * [16] 最接近的三数之和
 */

// @lc code=start
public partial class Solution
{
    public int ThreeSumClosest(int[] nums, int target)
    {
        int length = nums.Length;
        if (length < 3)
        {
            return 0;
        }
        Array.Sort(nums);
        int minDifference = int.MaxValue;
        for (int i = 0; i < length - 2; i++)
        {
            int num1 = nums[i];
            int outerTarget = target - num1;
            int left = i + 1;
            int right = length - 1;
            while (left < right)
            {
                int num2 = nums[left];
                int num3 = nums[right];
                int difference = outerTarget - num2 - num3;
                if (difference == 0)
                {
                    return target;
                }
                if (difference > 0)
                {
                    do
                    {
                        left++;
                    } while (left < right && nums[left] == nums[left - 1]);
                }
                if (difference < 0)
                {
                    do
                    {
                        right--;
                    } while (left < right && nums[right] == nums[right + 1]);
                }
                if (Math.Abs(minDifference) > Math.Abs(difference))
                {
                    minDifference = difference;
                }
            }
        }

        return target - minDifference;
    }
}
// @lc code=end

File: src/16.最接近的三数之和.rs
/*
 * @lc app=leetcode.cn id=16 lang=rust
 *
 * [16] 最接近的三数之和
 */

// @lc code=start
impl Solution {
    pub fn three_sum_closest(mut nums: Vec<i32>, target: i32) -> i32 {
        let length = nums.len();
        if length < 3 {
            return i32::MAX;
        }
        nums.sort();
        let mut min_diff = i32::MAX;
        for i in 0..length - 3 {
            let num1 = nums[i];
            let outer_target = target - num1;
            let mut left = i + 1;
            let mut right = length - 1;
            while left < right {
                let num2 = nums[left];
                let num3 = nums[right];
                let diff = outer_target - num2 - num3;
                if diff == 0 {
                    return target;
                }
                if diff > 0 {
                    loop {
                        left += 1;
                        if left >= right || nums[left] != nums[left - 1] {
                            break;
                        }
                    }
                }
                if diff < 0 {
                    loop {
                        right -= 1;
                        if left >= right || nums[right] != nums[right + 1] {
                            break;
                        }
                    }
                }
                if min_diff.abs() > diff.abs() {
                    min_diff = diff;
                }
            }
        }

        return target - min_diff;
    }
}

// @lc code=end
